### PR TITLE
Migrate BEEFY BLS crypto to  bls12-381 curve

### PR DESCRIPTION
**Pull request by drskalman**
We are definitely going to use BLS12-381 for BEEFY and it is hard coded in JAM's spec. This PR implements missing tests for bls12-381 crypto, migrate BEEFY BLS crypto to bls12-381 and adapt the BEEFY primitive tests accordingly.

### CODE REVIEW

File: substrate/primitives/core/src/bls.rs [hunk 1]
```diff
@@ -238,183 +238,194 @@ mod tests {
...
-		assert!(!Pair::verify(&signature, "Other message", &public));
-	}
-
-	#[test]
-	fn generate_with_phrase_recovery_possible() {
-		let (pair1, phrase, _) = Pair::generate_with_phrase(None);
-		let (pair2, _) = Pair::from_phrase(&phrase, None).unwrap();
-
-		assert_eq!(pair1.public(), pair2.public());
+        fn default_phrase_should_be_used_for_bls377() {
```
Comment (davxy) [inline substrate/primitives/core/src/bls.rs:254]: So if I understood correctly you want to share test code between bls12-381 and 377 right?
Looks fine

**Review by acatangiu — APPROVED** [review]
The BEEFY part looks good

**Review by davxy — COMMENTED** [review]
Mostly LGTM. Some considerations to discuss before approval:
1. Do you think we can remove the `bls-experimental` feature?
2. I see that in keystore and host functions (io) we have kept bls12-377 functionalities beside the new bls12-381. Maybe we can remove bls12-377 functions as far as they are not used/required/requested by anyone yet? I'd keep only the stuff in `crypto/core` for the moment (i.e. the lower level primitives)

File: substrate/primitives/consensus/beefy/src/witness.rs [hunk 2]
```diff
@@ -198,15 +198,15 @@ mod tests {
 			// from signed take a function as the aggregator 
 			TestBlsSignedCommitmentWitness::from_signed::<_, _>(signed, |sigs| {
 				// we are going to aggregate the signatures here
-				let mut aggregatedsigs: SignatureAggregatorAssumingPoP<TinyBLS377> =
+				let mut aggregatedsigs: SignatureAggregatorAssumingPoP<TinyBLS381> =
```
[the remark this answers is not in the document]
Comment (drskalman) [reply substrate/primitives/consensus/beefy/src/witness.rs:201]: I implemented such a thing during the Prague hackathon: https://github.com/w3f/polkadot-sdk/commit/53d90d3b745add6060b19651c7452be3fac0d14e#diff-8b079a4b5fd964b13e8f4c7da10f3a924d015f1dec8f4bad2059eb3ab20ca94fR246 I asked @AlistairStewart if I should port those efforts here. He disagreed. He said that there was a plan to publish signature accumulation (either Merkle root or BLS  aggregation) in BEEFY messages but that plan was scrapped. That is why you see "witness" only shows up in test. The current plan is that the aggregation should entirely happens by the relayer/prover. 

Having said that, I'm happy to do whatever gets this pull request approved >:-). I could remove the test entirely or if you really insists that the aggregation code shows up in the test or in the backend, I'd think Al won't object if the approval of the pull request hanging on this, his concern was mostly time spend on something we won't use I assume.

[the remark this answers is not in the document]
Comment (davxy) [reply substrate/primitives/consensus/beefy/src/witness.rs:201]: Was mostly a nitpick. What about the experimental feature and not exposing the bls12-377 stuff in keystore and sp-io?

[the remark this answers is not in the document]
Comment (drskalman) [reply substrate/primitives/consensus/beefy/src/witness.rs:201]: I'll remove bls12-377 from sp-io and application crypto and remove as much as needed to make it compile. I'd keep the experimental feature for now. I'd remove it in the next pull request when I implement proof of possession check, without that BLS aggregations are not secure.

**Comment by drskalman** [comment]
@davxy I removed bls12-377 from io and also I had to remove the application cryptos which were using the gen functions. The rest is not user facing and compiles. I assume it is Ok to stay there, in case of some future application (bridging to Celo for example)

**Comment by davxy** [comment]
@drskalman Looks good. Once the pipeline is fixed (I thing there are some clippy issues) we can merge

**Review by davxy — APPROVED** [review]

**Pull request merged**